feat(core): Adds selector support to dynamic component creation

SkyZeroZx opened the pull request:
#  Feature: Add Selector Support to Dynamic Component Creation

This PR implements the ability to specify a custom selector when dynamically creating components using `createComponent()`, addressing the limitation where developers couldn't control which selector from a multi-selector component definition gets used for the host element.

##  Motivation

Many UI libraries (like Angular Material) define components with multiple selectors to support different use cases:

```typescript
@Component({
  selector: `
    button[matButton], a[matButton], button[mat-button], button[mat-raised-button],
    button[mat-flat-button], button[mat-stroked-button], a[mat-button], a[mat-raised-button],
    a[mat-flat-button], a[mat-stroked-button]
  `,
  // ...
})
export class MatButton {}
```

Previously, `createComponent(MatButton)` would always use the first selector (`button[matButton]`), with no way to specify a different one like `a[mat-button]` for creating anchor-based buttons dynamically.

## Solution

Added an optional `selector` field to the `createComponent` options using a new function called `hostSelector`:

```typescript
createComponent(MatButton, { 
  environmentInjector, 
  selector: hostSelector('a[mat-button]')
});
```
 
## Features

- **Selector Validation**: Only selectors defined in the component are allowed.
- **Host Element Creation**: The host element matches the provided selector.
- **Error Handling**: Invalid selectors produce clear error messages.


##  Examples

### Basic Tag Selector
```typescript
createComponent(MyComponent, {
  environmentInjector,
  selector: hostSelector('custom-element')
});
```

### Complex Selector
```typescript
createComponent(MyComponent, {
  environmentInjector,
  selector: hostSelector('button.primary[type="submit"]')
});
```

### Angular Material Use Case
```typescript
const buttonRef = createComponent(MatButton, {
  environmentInjector,
  selector: hostSelector('a[mat-button]')
});
// Host element: <a mat-button></a>
```
## Error Handling

```typescript
createComponent(MyComponent, {
  environmentInjector,
  selector: hostSelector('invalid-selector')
});
// Throws: Invalid selector for component MyComponent. The selector 'invalid-selector' is not allowed.
```
 
 
 EDIT : 
 The implementation form was changed to a more functional one to make use of  Tree Shake, the code is only added when said function is used.

JeanMeche commented on packages/core/src/render3/component_parse_selector.ts:
```diff
@@ -0,0 +1,539 @@
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import {
```
The length of this file and the fact that the code is not removed by tree-shaking makes it a no-go. 

If we want to implement this feature, it should be at minimal cost / or that the cost is only payed by those who want to benefit from that feature.

SkyZeroZx replied:
Then I could migrate to a working version with three shake for its implementation, I will work on it

SkyZeroZx replied:
@JeanMeche I have already updated the code, the use of classes has been removed by migrating to a more Functional Programming version and I validated Tree Shake when the function is used or not.

SkyZeroZx commented:
> I don't think we either want to go toward a binding. This isn't a binding as the value can't and won't change over time.
> 
> Also the initial feature request was for `ViewContainRef.createComponent` (which creates a new node), not the standalone `createComponent` function (which uses an existing node as host).
> 
> The core feature we probably need to adapt is here
> 
> https://github.com/angular/angular/blob/67d4449db4cc23c5e770af27e185f40f2639f1ee/packages/core/src/render3/component_ref.ts#L186-L194

I think a better approach is to named this new feature `hostSelector` function, keeping the concept similar to a binding but in a separate file so it remains tree-shakeable. 

Additionally, to support this feature we should update both `createHostElement` and `createRootTView`, since the host element must match the parsed selector attributes.

In `createRootTView`, attributes and classes are currently extracted only from the first selector (to align with VE). For this feature, we need to ensure consistency with the host selector parsing so that the host element reflects the expected attributes.

```ts
function createRootTView(
  rootSelectorOrNode: any,
  componentDef: ComponentDef<unknown>,
  componentBindings: Binding[] | undefined,
  directives: (Type<unknown> | DirectiveWithBindings<unknown>)[] | undefined,
): TView {
  const tAttributes = rootSelectorOrNode
    ? ['ng-version', '0.0.0-PLACEHOLDER']
    : // Extract attributes and classes from the first selector only to match VE behavior.
      extractAttrsAndClassesFromSelector(componentDef.selectors[0]);
  ```
  
I also think we should add hostSelector support in both createComponent paths, since createComponent can also be attached dynamically in different parts of an Angular app. Or, if we want to keep the initial scope smaller, should this first be limited only to ViewContainerRef#createComponent?

Proposed scope: support in both ViewContainerRef#createComponent and the standalone createComponent API.
  
  Example with standalone createComponent i think is valid
   ```ts
       const buttonComponent = createComponent(AppIconButtonComponent, {
      environmentInjector: this.injector ,
      selector : hostSelector('button[appIconButton]')
    });
    document.body.appendChild(buttonComponent.location.nativeElement);
   ```
   Example wtih ViewContainerRef
   ```ts
          export class DummyComponent {
       
            private readonly viewContainerRef = inject(ViewContainerRef);
            addIcon() {
              const ref = this.viewContainerRef.createComponent(AppIconButtonComponent , { 
                        selector : hostSelector('button[appIconButton]')
              });
              ref.changeDetectorRef.detectChanges();
            }
          }

    });
   ```

SkyZeroZx commented:
@JeanMeche The pipeline can be run again. I was missing the linter and running the golden API.

SkyZeroZx commented:
We can re-run the pipeline, it failed due to a lack of golden symbols, my WSL has been having some strange issues with PNPM

SkyZeroZx committed:
feat(core): Adds selector support to dynamic component creation

Enables specifying a host selector when dynamically creating components,allowing for more precise control over the host element's tag and attributes. Validates selectors against component definitions to prevent invalid usage.

thePunderWoman commented:
I think we're going to close this for now as we're not sure this is the direction we want to go with dynamic component creation. Thanks though!
